pkg/ccl/sqlproxyccl/forwarder.go

```diff
@@ -177,3 +158,304 @@ func wrapServerToClientError(err error) error {
+	// message). If the suspend request failed, an error will be returned. The
+	// caller is safe to retry again.
+	suspend() error
+	// waitUntilSuspended blocks until the processor has been suspended. If
+	// suspend was not called before, this may block forever.
+	waitUntilSuspended()
+}
+
+// requestProcessor handles the communication from the client to the server.
+type requestProcessor struct {
+	f  *forwarder
+	mu struct {
+		syncutil.Mutex
+		cond       *sync.Cond
+		resumed    bool
+		inPeek     bool
+		suspendReq bool // Indicates that a suspend has been requested.
+	}
+	wg sync.WaitGroup
+}
+
+var _ processor = &requestProcessor{}
+
+func (p *requestProcessor) lock() {
+	p.mu.Lock()
+	if p.mu.cond == nil {
+		p.mu.cond = sync.NewCond(&p.mu)
+	}
+}
+
+func (p *requestProcessor) unlock() { p.mu.Unlock() }
+
+func (p *requestProcessor) resume() error {
+	p.wg.Add(1)
+	defer p.wg.Done()
+
+	// Has context been cancelled?
+	if p.f.ctx.Err() != nil {
+		return p.f.ctx.Err()
+	}
+
+	p.lock()
+	// Already resumed - don't start another one.
+	if p.mu.resumed {
+		p.unlock()
+		return errProcessorResumed
+	}
+	p.mu.resumed = true
+	p.unlock()
+	defer func() {
+		p.lock()
+		defer p.unlock()
+		p.mu.resumed = false
+	}()
+
+	for p.f.ctx.Err() == nil {
+		// inPeek has to be false before this because the resumed field guards
+		// against concurrent runs.
+		p.lock()
+		// Suspend has been requested. Suspend now before blocking.
+		if p.mu.suspendReq {
+			p.mu.suspendReq = false
+			p.unlock()
+			return nil
+		}
+		p.mu.inPeek = true
+		if dErr := p.f.clientConn.SetReadDeadline(time.Time{}); dErr != nil {
+			p.unlock()
+			return dErr
+		}
+		p.unlock()
+
+		// Always peek the message to ensure that we're blocked on reading the
+		// header, rather than when forwarding.
+		_, _, err := p.f.clientConn.PeekMsg()
+
+		p.lock()
+		suspend := p.mu.suspendReq
+		p.mu.suspendReq = false
+		p.mu.inPeek = false
+		if dErr := p.f.clientConn.SetReadDeadline(time.Time{}); dErr != nil {
+			p.unlock()
+			p.mu.cond.Broadcast()
+			return dErr
+		}
+		p.unlock()
+		p.mu.cond.Broadcast()
+
+		// If suspend was requested, there are two cases where we will
+		// terminate:
+		//   1. err == nil, where we managed to read a header. In that case,
+		//      suspension gets priority.
+		//   2. err != nil, where the error was due to a timeout. Connection
+		//      is likely idle here.
+		if err != nil {
+			if netErr := (net.Error)(nil); errors.As(err, &netErr) && suspend && netErr.Timeout() {
+				// Do nothing.
+				err = nil
+			} else {
+				err = errors.Wrap(err, "peeking message")
+			}
+		}
+		if err != nil || suspend {
+			return err
+		}
+
+		if _, err := p.f.clientConn.ForwardMsg(p.f.serverConn); err != nil {
+			return errors.Wrap(err, "forwarding message")
+		}
+	}
+	return p.f.ctx.Err()
+}
+
+func (p *requestProcessor) suspend() error {
+	p.lock()
+	defer p.unlock()
+	// Processor has not been resumed.
+	if !p.mu.resumed {
```

Comment:
nit: we can simplify suspend if it waits for resumed == false.

E.x:

```
p.lock()
defer p.unlock()
p.mu.suspendReq = true
for p.mu.resumed {
  if p.mu.inPeek {
    p.f.clientConn.SetReadDeadline(aLongTimeAgo)
  }
  p.mu.cond.Wait()
}
return nil
```

Reply:
Hm, I guess doing this also means that we can remove waitUntilSuspended. I'll look into this, and see how this ties back to the connection migration.